### PR TITLE
Publishing option for signing a message with a custom private key

### PR DESCRIPTION
**Pull request by iulianpascalau**
- added publishing option for signing a message with a custom private key

### CODE REVIEW

**Review by vyzo — COMMENTED** [review]
This looks fine, but would it make more sense to introduce a topic option `WithPublisher` and using that?

File: topic.go [hunk 1]
```diff
@@ -286,6 +293,54 @@ func (t *Topic) Publish(ctx context.Context, data []byte, opts ...PubOpt) error
 	return t.p.val.PushLocal(&Message{m, "", t.p.host.ID(), nil})
 }
 
+// PublishWithSk publishes data to topic using the provided secret key and generated peer ID. Providing also the peer ID
+// (which can be cached) will improve the performance as it does not require the generation of the public key
+// and the resulting peer ID.
+// This method can be used as to be able to publish messages on the network without having a "real", connectable host.
+func (t *Topic) PublishWithSk(ctx context.Context, data []byte, signKey crypto.PrivKey, pid peer.ID, opts ...PubOpt) error {
```
Comment (vyzo) [inline topic.go]: let's not abbreviate the Sk and say `PublishWithKey` or `PublishWithSecretKey`.

Comment (iulianpascalau) [reply topic.go:300]: > This looks fine, but would it make more sense to introduce a topic option `WithPublisher` and using that?

I will explore that. I was on the verge of pushing a new commit for code duplication removal.

**Comment by iulianpascalau** [comment]
> This looks fine, but would it make more sense to introduce a topic option `WithPublisher` and using that?

I will explore that. I was on the verge of pushing a new commit for code duplication removal.

File: topic.go (hunk 1 above)
Comment (iulianpascalau) [reply topic.go]: will refactor

File: topic.go [hunk 2]
```diff
@@ -216,27 +227,40 @@ func (t *Topic) Publish(ctx context.Context, data []byte, opts ...PubOpt) error
...
+		}
+	}
+
 	m := &pb.Message{
 		Data:  data,
 		Topic: &t.topic,
 		From:  nil,
 		Seqno: nil,
 	}
 	if t.p.signID != "" {
```
Comment (vyzo) [inline topic.go]: BUG: this should check the local pid variable.

File: topic.go [hunk 3]
```diff
@@ -216,27 +227,40 @@ func (t *Topic) Publish(ctx context.Context, data []byte, opts ...PubOpt) error
...
-		m.From = []byte(t.p.signID)
+		m.From = []byte(pid)
 		m.Seqno = t.p.nextSeqno()
 	}
 	if t.p.signKey != nil {
```
Comment (vyzo) [inline topic.go]: BUG: this should check the local key variable

**Review by vyzo — CHANGES_REQUESTED** [review]
there are a couple of bugs, but otherwise looks good.

File: topic.go (hunk 3 above)
Comment (iulianpascalau) [reply topic.go]: right, fixed. Thanks

Comment (iulianpascalau) [reply topic.go:261]: right, fixed. Thanks

File: topic.go (hunk 2 above)
Comment (iulianpascalau) [reply topic.go]: right, fixed. Thanks

File: topic_test.go [hunk 4]
```diff
@@ -920,3 +921,101 @@ func TestWithTopicMsgIdFunction(t *testing.T) {
 		t.Fatal("msg ids are equal")
 	}
 }
+
+func TestTopic_PublishWithKeyInvalidParameters(t *testing.T) {
```
Comment (vyzo) [inline topic_test.go]: can we remove the underscore?

File: topic_test.go [hunk 5]
```diff
@@ -920,3 +921,101 @@ func TestWithTopicMsgIdFunction(t *testing.T) {
...
+	t.Run("empty peer ID should error", func(t *testing.T) {
+		withVirtualKey := WithSecretKeyAndPeerId(virtualPeer.PrivKey, "")
+		err := topics[0].Publish(ctx, []byte("buff"), withVirtualKey)
+		if err != ErrEmptyPeerID {
+			t.Fatal("error should have been of type errEmptyPeerID")
+		}
+	})
+}
+
+func TestTopicRelay_PublishWithKey(t *testing.T) {
```
Comment (vyzo) [inline topic_test.go]: here too

**Review by vyzo — COMMENTED** [review]
just some final nits and it is ready to merge.

File: topic_test.go (hunk 4 above)
Comment (iulianpascalau) [reply topic_test.go]: sure

Comment (iulianpascalau) [reply topic_test.go]: this is the Elrond pattern used in tests :)

Comment (iulianpascalau) [reply topic_test.go]: done

File: topic_test.go (hunk 5 above)
Comment (iulianpascalau) [reply topic_test.go]: done

**Review by vyzo — APPROVED** [review]
thank you!

**Pull request merged**